bindings/rust/s2n-tls/src/callbacks.rs

```diff
@@ -118,30 +124,128 @@ pub(crate) fn trigger_async_callback<T: 'static + AsyncCallback>(
+        // this should NEVER be a `&mut` since its would be possible to
+        // mutate the Config on the Connection. Since one Config can be
+        // used for multiple Connections that would be undefined-behavior
+        &self,
+        conn: &mut Connection,
+    ) -> Poll<Result<(), Error>>;
+}
+
+/// The Future associated with the async [`ClientHelloCallback`] and stored
+/// on the connection.
+///
+/// The calling application can provide an instance of [`AsyncClientHelloFuture`]
+/// when implementing the [`ClientHelloCallback`] if it wants to run an
+/// asynchronous operation (disk read, network call). The application can return
+/// an error, [`Err(error::ApplicationError)`], to indicate failure and cancel the
+/// connection.
+///
+/// [`ConfigResolver`] should be used if the application wants to set a new
+/// [`Config`] on the connection.
+pub trait AsyncClientHelloFuture {
+    fn poll_client_hello(
+        self: Pin<&mut Self>,
+        connection: &mut Connection,
+        ctx: &mut core::task::Context,
+    ) -> Poll<Result<(), Error>>;
+}
+
+pin_project! {
+/// An implementation of [`AsyncClientHelloFuture`] which resolves the provided
+/// future and sets the config on the [`connection::Connection`].
+pub struct ConfigResolver<F: Future<Output = Result<Config, Error>>> {
+    #[pin]
+    fut: F,
+}
+}
+
+impl<F: Future<Output = Result<Config, Error>>> ConfigResolver<F> {
+    pub fn new(fut: F) -> Self {
+        ConfigResolver { fut }
+    }
+}
+
+impl<F: Future<Output = Result<Config, Error>>> AsyncClientHelloFuture for ConfigResolver<F> {
+    fn poll_client_hello(
+        self: Pin<&mut Self>,
+        connection: &mut Connection,
+        ctx: &mut core::task::Context,
+    ) -> Poll<Result<(), Error>> {
+        let this = self.project();
+        let config = match this.fut.poll(ctx) {
+            Poll::Ready(config) => config?,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        connection.set_config(config)?;
+
+        Poll::Ready(Ok(()))
+    }
 }
 
 /// A trait for the callback executed after parsing the TLS Client Hello.
 ///
 /// Use in conjunction with
 /// [config::Builder::set_client_hello_callback](`crate::config::Builder::set_client_hello_callback()`).
 pub trait ClientHelloCallback {
-    fn poll_client_hello(&self, connection: &mut Connection) -> Poll<Result<(), Error>>;
+    /// The application can return a `None` to resolve the client_hello_callback
+    /// synchronously or return a `Some(AsyncClientHelloFuture)` if it wants to
+    /// run some asynchronous task beforing resolving the callback.
+    ///
+    /// [`ConfigResolver`], which implements [`AsyncClientHelloFuture`] can be
+    /// returned if the application wants to set a new [`Config`] on the connection.
+    ///
+    /// If the server_name is used to configure the connection then the application
+    /// should call [`connection::Connection::server_name_extension_used()`].
+    fn on_client_hello(
```

Comment:
i think it's reasonable for this method to return an `Result<..., Error>`. Otherwise, we're forcing the implementation to return a allocated future that immediately returns